### PR TITLE
net/sock_dtls: add public key verification

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description

This adds a pseudomodule to enable public key verification of peers. When `sock_dtls_verify_public_key` is enabled the public key sent by the peer is checked against the registered keys on the corresponding sock. The implementation for tinydtls is added, as well as the corresponding changes to use it on the `dtls-sock` example.

### Testing procedure
Run `examples/dtls-sock` with ECC enabled, the public key should be verified. You can try removing the server's public key from the client, the connection should be aborted.

### Issues/PRs references
Loosely related to #19838

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
I'm wondering whether this feature should actually be opt-out instead.

**Review by Teufelchen1 — CHANGES_REQUESTED** [review]
Some nits. Manually and successfully tested on native.

I found the code in the example to be confusing due to naming of the key variables and lack of comments. If you want to go the extra mile and and a bit of :sparkles: that would be neat but totally optional for this PR.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: Done, also I switched to use `DTLS_EC_KEY_SIZE` instead.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: Removed

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: Changed

File: examples/dtls-sock/dtls-client.c
Comment (leandrolanzieri) [reply examples/dtls-sock/dtls-client.c:203]: > I was always very confused about what's the idea behind credman.

Yeah, I know what you mean...

> Shouldn't a credential be associated with a host? So what does it mean to add multiple credentials to a socket?

Not necessarily. I mean, you can already see the example of the server with more than one pair of keys. IMO one of the greatest problems of this module is that a credential is a single "private key" and 0 or more "client keys". Like I commented in the example, the way it currently works looks more like a keyring of client keys and your own.

> So what's the advantage of the credential IDs ('tags') as opposed to just passing pointers?

In key management using tags is not uncommon, this is crucial for instance when integrating secure elements or other backends where the key is directly not reachable, but a tag is passed to the crypto backend. I'm not sure why this was the case for credman, as it clearly doesn't support any other backend than a buffer in memory.

IMO we are in need of a rethinking of the credentials management in RIOT. We would really benefit from:
- not only DTLS credentials (of course)
- possibility of multiple backends including secure elements
- more types of credentials
- ability to indicate scope and capabilities of the credentials

**Comment by benpicco** [comment]
This needs a rebase

**Comment by leandrolanzieri** [comment]
Rebased. I also modified a bit the example: the server only stores one client public key, but has two key pairs. It allows the user to select which one to use on the handshake. As the client knows both public keys form the server,  any will work. I think this showcases a bit better the usage of the callbacks and multiple credentials. I also updated the docs including this and the new command.

File: examples/dtls-sock/dtls-client.c
Comment (Teufelchen1) [reply examples/dtls-sock/dtls-client.c:193]: Hey @Einhornhool! Take a look at this comment please :tiger2: 
(Also, maybe this comment should be an issue instead?)

**Comment by Teufelchen1** [comment]
This is great! Way easier to understand!
Could you fix the format issues in the `*credentials.h` files?
Don't forget to squash!

**Comment by leandrolanzieri** [comment]
Fixed and squashed!

File: examples/dtls-sock/dtls-client.c
Comment (benpicco) [inline examples/dtls-sock/dtls-client.c:139]: Why move away from the `sock_dtls_establish_session()` helper and open-code the handshake again?

Comment (leandrolanzieri) [reply examples/dtls-sock/dtls-client.c:139]: Because it doesn't allow me to register the extra credentials before initiating the handshake.

Comment (leandrolanzieri) [reply examples/dtls-sock/dtls-client.c:139]: Same with the callbacks

Comment (benpicco) [reply examples/dtls-sock/dtls-client.c:139]: tbh I didn't quite understand the requirement when adding that helper, so if you think the API can be improved that would be great - leaving the handshake to the user seemed like a bad example.

Comment (Teufelchen1) [reply examples/dtls-sock/dtls-client.c:139]: Hm I don't feel like this example is bad. It's not overly complicated or lengthy, the reader should get along well.
While a single helper function is nice and clean, it is clear the existing one is insufficient for this particular task. Adjusting that helper would mess with an API that is out of scope for this PR.

* Can we resolve this comment on these terms? 
* Do you think the API needs a rework? If so, we should open an issue for that.
* Makes me wonder, should we integrate credman closer with dtls?

Comment (benpicco) [reply examples/dtls-sock/dtls-client.c:139]: Sure, this shouldn't block this PR, I just took the opportunity to try to get some open questions about credman / DTLS answered :sweat_smile: 

I'd very much welcome an overhaul of credman as it seems still a bit unclear what it actually is and what's it's purpose.

Maybe it would also fit into PSA crypto?

**Review by Teufelchen1 — APPROVED** [review]

File: sys/include/net/sock/dtls.h
Comment (benpicco) [inline sys/include/net/sock/dtls.h:514]: This is only true for ECDSA, right? 
AFAIK with the default `CREDMAN_TYPE_PSK` there are no public/private keys, only symmetric encryption.

File: examples/dtls-sock/dtls_client_credentials.h
Comment (benpicco) [inline examples/dtls-sock/dtls_client_credentials.h:45]: Would the hint be something like a hostname / node ID or serial number?

File: sys/include/net/sock/dtls.h
Comment (benpicco) [reply sys/include/net/sock/dtls.h:514]: I don't think that's obvious (it certainly was not for me) that `CREDMAN_TYPE_PSK` is symmetric only whereas `CREDMAN_TYPE_ECDSA` will do a Diffie-Hellman handshake (and use symmetric crypto afterwards)

File: examples/dtls-sock/dtls_client_credentials.h
Comment (benpicco) [reply examples/dtls-sock/dtls_client_credentials.h:45]: Sure, but since this is an example I think some comment of what this would be in a real-world use case would be valuable.

File: examples/dtls-sock/dtls-client.c
Comment (Einhornhool) [reply examples/dtls-sock/dtls-client.c:193]: Hmm most of the points in Leandros credentials requirement list could be handled by the PSA Crypto key management. It might be worth looking into patching tinydtls to use PSA for crypto operations and credential storage, instead of implementing the same features twice.

File: examples/dtls-sock/dtls_client_credentials.h
Comment (leandrolanzieri) [reply examples/dtls-sock/dtls_client_credentials.h:45]: > something like a hostname / node ID or serial number

This would somehow imply that the key is only to be used with a single client, this is not necessarily the case. Searching for examples I found this summary: https://stackoverflow.com/a/8753393/5633371 seems that many implementations send encoded information on how to derive the keys.

I think this hint is fine for the didactic purpose of the example: It's easily readable in code and when sniffing the handshake, and it describes its intention.

**Comment by leandrolanzieri** [comment]
Ping?

File: examples/dtls-sock/Makefile
Comment (benpicco) [inline examples/dtls-sock/Makefile:23]: Does this do anything with the default `CONFIG_DTLS_PSK`?

**Review by benpicco — COMMENTED** [review]
Thank you for the ping.

@Teufelchen1 I see you approved this, any reason not to press the merge button?

File: examples/dtls-sock/Makefile
Comment (Teufelchen1) [reply examples/dtls-sock/Makefile:23]: No it does not. I could be guarded away - on the other hand, the impact is very minor and can be seen as test case: psk still works even when ecc verification is active?

Comment (Teufelchen1) [reply examples/dtls-sock/Makefile:23]: Okay, summary:
- Even if `sock_dtls_verify_public_key` is used with `CONFIG_DTLS_PSK` the verification part is not being build until `ecc` is selected
- Verification should be the default anyway, also noted in leandros first comment to this pr "I'm wondering whether this feature should actually be opt-out instead."
- For user of this example it is easier to switch between `psk/ecc` rather than bothering them with activating verification
- It could get guarded in the makefile but the benefit is minimal with the downside of making the Makefile less approachable

We believe it can stay the way it is, @benpicco?

Comment (Teufelchen1) [inline examples/dtls-sock/Makefile:23]:
```suggestion
# When using asymmetric cryptography, e.g. ecc, 
# this verifies the public key against a known set.
# It does not affect symmetric crypto e.g. PSK.
USEMODULE += sock_dtls_verify_public_key
```

Everyone happy with this addition?

**Comment by leandrolanzieri** [comment]
Thanks for the review!

**Pull request merged**
